rfcs/syntax-nil-navigation-operator.md

```diff
@@ -0,0 +1,116 @@
+# nil-navigation postfix operator (?)
+
+## Summary
+
+Introduce syntax to navigate through `nil` values.
+
+## Motivation
+
+nil values are very common in Lua, and take care to prevent runtime errors. 
+
+Currently, attempting to index `dog.name` while caring for `dog` being nil requires some form of the following:
+
+```lua
+local dogName = nil
+if dog ~= nil then
+    dogName = dog.name
+end
+```
+
+...or the unusual to read...
+
+```lua
+local dogName = dog and dog.name
+```
+
+...which will return `false` if `dog` is `false`, instead of throwing an error because of the index of `false.name`.
+
+This trick gets worse in the case of calling methods. For example, let's suppose we wanted to call `dog.fetch()`, while `dog` is still potentially nil.
+
+The one-line `and` trick will no longer work (or will get less readable as you try to shoe-horn this in), so we must:
+
+```lua
+if dog ~= nil then
+    dog.fetch()
+end
+```
+
+But this gets even worse when it comes to chained indexing. Let's suppose we wanted to run `dog.owner.handshake()`, while `dog` can be nil and `owner` can be nil.
+
+```lua
+if dog ~= nil and dog.owner ~= nil then
+    dog.owner.handshake()
+end
+```
+
+...which gets even worse in the context of calling this function in, say, another function:
+
+```lua
+-- Oops! dog and dog.owner can be nil
+logDogName(getLogger(), dog.name, dog.owner:getDisplayName())
+
+-- In order to preserve this order (assuming argument execution order mattered)...
+local logger = getLogger()
+local name = dog and dog.name
+local displayName
+
+if dog.owner ~= nil then
+    displayName = dog.owner:getDisplayName()
+end
+
+logDogName(logger, name, displayName)
+```
+
+## Design
+
+The nil-navigation operator will make all of these smooth, by supporting `x?.y` and similar indexing operators. `dog?.name` would resolve to `nil` if `dog` was nil, or the name otherwise. `owner?.handshake()` would only call `handshake` if `owner` is not nil.
+
+The long example would turn into:
+
+```lua
+logDogName(getLogger(), dog?.name, dog?.owner?:getDisplayName())
+```
+
+Failing the nil-safety check early would make the entire expression nil, for instance `dog?.body.legs` would resolve to `nil` if `dog` is nil, rather than resolve `dog?.body` into nil, then turning into `nil.legs`.
+
+Because functions can also be nil, the nil-navigation operator will also support `x?()`, which will only call `x` if it is not nil.
+
+The list of valid operators to follow the nil-navigation operator would be:
+
+```lua
+dog?.name
+dog?.getName()
+dog?:getName()
+dogs?[1]

```

Comment:
Is this and the call below compelling enough to introduce? Trying to gauge if it might be true that `?.` is 80% of the value, `?:` is the remaining 15% and the rest is mostly noise?

Reply:
Likely not compelling enough for an initial release, especially since you're unlikely to chain them as much.

Reply:
I moved it to alternatives for now, noting that they are not out of scope but are not necessary.